File: app/src/main/java/cn/intret/app/picgo/ui/main/MainContract.java
package cn.intret.app.picgo.ui.main;

import android.support.annotation.NonNull;
import android.support.annotation.StringRes;

import java.io.File;
import java.util.List;

import cn.intret.app.picgo.model.image.FolderModel;
import cn.intret.app.picgo.model.image.MediaFile;
import cn.intret.app.picgo.model.user.SortOrder;
import cn.intret.app.picgo.model.user.SortWay;
import cn.intret.app.picgo.model.user.UserInitialPreferences;

/**
 * Created by intret on 2018/3/13.
 */

public class MainContract {

    enum LoadImageListPurpose {
        RefreshDetailList,
        RefreshDefaultList
    }

    interface View {
        void onLoadedUserInitialPreferences(UserInitialPreferences userInitialPreferences);

        void onErrorMessage(@StringRes int msg);

        void onLoadedFolderModel(@NonNull FolderModel folderModel);

        void onLoadedFolderModel(FolderModel folderModel, boolean diff);

        void onLoadFolderModelFailed(Throwable throwable);

        void onDeletedDirectory(File dir);

        void onDeleteDirectoryFailed(File dir, Throwable throwable);

        void onDeleteNotEmptyDirectoryFailed(File dir, Integer fileCount);

        void onDiffDetailMediaFiles(File dir, List<MediaFile> mediaFiles, boolean hideRefreshControl, LoadImageListPurpose purpose);

        void onDiffLoadDetailMediaFileFailed(File dir, boolean hideRefreshControl, LoadImageListPurpose purpose);

        void onDiffLoadedDefaultMediaFiles(File dir, List<MediaFile> mediaFiles, boolean hideRefreshControl);

        void onDiffLoadDefaultMediaFileFailed(File dir, boolean hideRefreshControl);

    }

    interface Presenter {
        void loadInitialPreference();

        void removeDirectory(File dir, boolean forceDelete);

        void loadFolderList(boolean fromCacheFirst);

        void loadFolderList(boolean fromCacheFirst, boolean diff);

        void diffLoadMediaFileList(File dir, boolean fromCacheFirst, SortWay sortWay, SortOrder sortOrder, boolean hideRefreshControl, LoadImageListPurpose purpose);

        void diffLoadDefaultMediaFileList(File dir, boolean fromCacheFirst, SortWay sortWay, SortOrder sortOrder, boolean hideRefreshControl);

        void updateFolderListItemThumbnailList(File directory);
    }
}

File: app/src/main/java/cn/intret/app/picgo/screens/main/MainContract.kt
package cn.intret.app.picgo.screens.main

import android.support.annotation.StringRes
import cn.intret.app.picgo.model.image.data.FolderModel
import cn.intret.app.picgo.model.image.data.MediaFile
import cn.intret.app.picgo.model.user.data.SortOrder
import cn.intret.app.picgo.model.user.data.SortWay
import cn.intret.app.picgo.model.user.data.UserInitialPreferences
import java.io.File

/**
 * Created by intret on 2018/3/13.
 */

class MainContract {

    enum class LoadImageListPurpose {
        RefreshDetailList,
        RefreshDefaultList
    }

    interface View {
        fun onLoadedUserInitialPreferences(userInitialPreferences: UserInitialPreferences)

        fun onErrorMessage(@StringRes msg: Int)

        fun onLoadedFolderModel(folderModel: FolderModel)

        fun onLoadedFolderModel(folderModel: FolderModel, diff: Boolean)

        fun onLoadFolderModelFailed(throwable: Throwable)

        fun onDeletedDirectory(dir: File)

        fun onDeleteDirectoryFailed(dir: File, throwable: Throwable)

        fun onDeleteNotEmptyDirectoryFailed(dir: File, fileCount: Int?)

        fun onDiffDetailMediaFiles(dir: File, mediaFiles: List<MediaFile>, hideRefreshControl: Boolean, purpose: LoadImageListPurpose)

        fun onDiffLoadDetailMediaFileFailed(dir: File, hideRefreshControl: Boolean, purpose: LoadImageListPurpose)

        fun onDiffLoadedDefaultMediaFiles(dir: File, mediaFiles: List<MediaFile>, hideRefreshControl: Boolean)

        fun onDiffLoadDefaultMediaFileFailed(dir: File, hideRefreshControl: Boolean)

    }

    internal interface Presenter {
        fun loadInitialPreference()

        fun removeDirectory(dir: File, forceDelete: Boolean)

        fun loadFolderList(fromCacheFirst: Boolean)

        fun loadFolderList(fromCacheFirst: Boolean, diff: Boolean)

        fun diffLoadMediaFileList(dir: File, fromCacheFirst: Boolean, sortWay: SortWay, sortOrder: SortOrder, hideRefreshControl: Boolean, purpose: LoadImageListPurpose)

        fun diffLoadDefaultMediaFileList(dir: File, fromCacheFirst: Boolean, sortWay: SortWay, sortOrder: SortOrder, hideRefreshControl: Boolean)

        fun updateFolderListItemThumbnailList(directory: File)
    }
}
